File: dbflow/src/main/java/com/raizlabs/android/dbflow/sql/language/Select.java
package com.raizlabs.android.dbflow.sql.language;

import android.support.annotation.NonNull;

import com.raizlabs.android.dbflow.sql.Query;
import com.raizlabs.android.dbflow.sql.QueryBuilder;
import com.raizlabs.android.dbflow.sql.language.property.IProperty;
import com.raizlabs.android.dbflow.sql.language.property.Property;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Description: A SQL SELECT statement generator. It generates the SELECT part of the statement.
 */
public class Select implements Query {

    /**
     * Default does not include the qualifier
     */
    public static final int NONE = -1;
    /**
     * The select qualifier to append to the SELECT statement
     */
    private int mSelectQualifier = NONE;
    /**
     * SELECT DISTINCT call
     */
    public static final int DISTINCT = 0;
    /**
     * SELECT ALL call
     */
    public static final int ALL = 1;

    private final List<IProperty> propertyList = new ArrayList<>();

    /**
     * Creates this instance with the specified columns from the specified {@link com.raizlabs.android.dbflow.config.FlowManager}
     *
     * @param properties The properties to select from.
     */
    public Select(IProperty... properties) {
        Collections.addAll(propertyList, properties);

        if (propertyList.isEmpty()) {
            propertyList.add(Property.Companion.getALL_PROPERTY());
        }
    }

    /**
     * Passes this statement to the {@link From}
     *
     * @param table    The model table to run this query on
     * @param <TModel> The class that implements {@link com.raizlabs.android.dbflow.structure.Model}
     * @return the From part of this query
     */
    @NonNull
    public <TModel> From<TModel> from(@NonNull Class<TModel> table) {
        return new From<>(this, table);
    }

    /**
     * appends {@link #DISTINCT} to the query
     *
     * @return
     */
    @NonNull
    public Select distinct() {
        return selectQualifier(DISTINCT);
    }

    @NonNull
    public String toString() {
        return getQuery();
    }

    @Override
    public String getQuery() {
        QueryBuilder queryBuilder = new QueryBuilder("SELECT ");

        if (mSelectQualifier != NONE) {
            if (mSelectQualifier == DISTINCT) {
                queryBuilder.append("DISTINCT");
            } else if (mSelectQualifier == ALL) {
                queryBuilder.append("ALL");
            }
            queryBuilder.appendSpace();
        }

        queryBuilder.append(QueryBuilder.join(",", propertyList));
        queryBuilder.appendSpace();
        return queryBuilder.getQuery();
    }


    /**
     * Helper method to pick the correct qualifier for a SELECT query
     *
     * @param qualifierInt Can be {@link #ALL}, {@link #NONE}, or {@link #DISTINCT}
     * @return
     */
    private Select selectQualifier(int qualifierInt) {
        mSelectQualifier = qualifierInt;
        return this;
    }
}

File: dbflow/src/main/java/com/raizlabs/android/dbflow/sql/language/Select.kt
package com.raizlabs.android.dbflow.sql.language

import com.raizlabs.android.dbflow.sql.Query
import com.raizlabs.android.dbflow.sql.language.property.IProperty
import com.raizlabs.android.dbflow.sql.language.property.Property
import kotlin.reflect.KClass

/**
 * Description: A SQL SELECT statement generator. It generates the SELECT part of the statement.
 */
class Select
/**
 * Creates this instance with the specified columns from the specified [com.raizlabs.android.dbflow.config.FlowManager]
 *
 * @param properties The properties to select from.
 */
(vararg properties: IProperty<*>) : Query {
    /**
     * The select qualifier to append to the SELECT statement
     */
    private var selectQualifier = NONE

    private val propertyList = arrayListOf<IProperty<*>>()

    override val query: String
        get() {
            val queryBuilder = StringBuilder("SELECT ")

            if (selectQualifier != NONE) {
                if (selectQualifier == DISTINCT) {
                    queryBuilder.append("DISTINCT")
                } else if (selectQualifier == ALL) {
                    queryBuilder.append("ALL")
                }
                queryBuilder.append(" ")
            }

            queryBuilder.append(propertyList.joinToString(separator = ","))
            queryBuilder.append(" ")
            return queryBuilder.toString()
        }

    init {
        propertyList.addAll(properties.toList())
        if (propertyList.isEmpty()) {
            propertyList.add(Property.ALL_PROPERTY)
        }
    }

    /**
     * Passes this statement to the [From]
     *
     * @param table    The model table to run this query on
     * @param [T] The class that implements [com.raizlabs.android.dbflow.structure.Model]
     * @return the From part of this query
     */
    fun <T : Any> from(table: Class<T>): From<T> = From(this, table)

    inline fun <reified T : Any> from() = from(T::class.java)

    infix fun <T : Any> from(table: KClass<T>) = from(table.java)

    /**
     * appends [.DISTINCT] to the query
     *
     * @return
     */
    fun distinct(): Select = selectQualifier(DISTINCT)

    override fun toString(): String = query


    /**
     * Helper method to pick the correct qualifier for a SELECT query
     *
     * @param qualifierInt Can be [.ALL], [.NONE], or [.DISTINCT]
     * @return
     */
    private fun selectQualifier(qualifierInt: Int) = apply {
        selectQualifier = qualifierInt
    }

    companion object {

        /**
         * Default does not include the qualifier
         */
        @JvmField
        val NONE = -1
        /**
         * SELECT DISTINCT call
         */
        @JvmField
        val DISTINCT = 0
        /**
         * SELECT ALL call
         */
        @JvmField
        val ALL = 1
    }
}

inline val select: Select
    get() = SQLite.select()
